### DISCUSSION TITLE
Reading telemtry from USB Serial Port

### DISCUSSION THREAD

**Discussion by bruno-luca** [yamcs/yamcs#676] [Q&A] [answered]
Hi!
First of all, thanks to this community for helping me through my developing process!

I have a Arduino that need to send data to Yamcs via **`USB serial port`**. I was looking inside the `yamcs server manual`  but i couldn't find a proper way to read data from usb serial. 
Is there a way to do it in Yamcs? If not, what could i do?

**Comment by xpromache** [comment]
There is something called TSE which can be used to connect to
test equipment implementing SCPI protocol.

But for your Arduino I think the easiest is to write a program in python
(or something else) that reads data via the serial port and pushes it to
Yamcs via network.

**Reply by bruno-luca** [reply]
I also looked at the DataLink list and read about FilePollingDataLink. I could create my own class based on AbstractTmDataLink and Similar to FilePollingDataLink  and implement a function that read from the usb serial port. Could it be possibile?

**Reply by xpromache** [reply]
yes, indeed, you can just copy the FilePollingTmDataLink. 

Maybe we should add an option in that class (incomingFile as alternative to the current incomingDir) rather than making another link.

If you feel up to it, I will gladly review and accept a pull request.

**Comment by merose** [comment]
Since it is a serial connection, you can get out of sync with the sender.
You may wish to add code to make sure you have a complete packet, so you
may want to look for a start sequence. That might complicate
FilePollingDataLink unnecessarily.

Mark

On Wed, May 18, 2022 at 4:35 AM Nicolae Mihalache ***@***.***>
wrote:

> yes, indeed, you can just copy the FilePollingTmDataLink.
>
> Maybe we should add an option in that class (incomingFile as alternative
> to the current incomingDir) rather than making another link.
>
> If you feel up to it, I will gladly review and accept a pull request.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yamcs/yamcs/discussions/676#discussioncomment-2775160>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMSVZKMD6PFXP6IEUI4SPLVKTIXJANCNFSM5WEBBLHQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Mark Rose
***@***.***
925.922.7263

**Comment by xpromache** [comment]
Marc is right, we should have a separate class dedicated to the serial links. Then we can also have some options for setting line speed, parity, etc.

**Comment by bruno-luca** [comment]
Thanks for your suggestions. I'm going to try a bounch of differents approaches. I'll let you know when i have something working

**Comment by bruno-luca** [comment, marked as answer]
Hello every one. I ended up with a solution that in my opinion is the one that bast fit my case. 
I saw the [service list](https://docs.yamcs.org/yamcs-server-manual/services/) on the documentation and looked at `ProcessRunner`, so i defined in the services file configuration the instruction to launch my own process, and it works. 

But i have two problems with `ProcessRunner ` service:
1. If my process launch a python script every print inside it aren't displayed on the terminal. The only way i could get them was by `ctrl+c` killing yamcs in the terminal, after that i get everything the script should have printed during its running
2. In the definition of [ProcessRunner](https://docs.yamcs.org/yamcs-server-manual/services/global/process-runner/) `restart` argument should set the behaviour of the program on some occasion, but if i define the service as follow: 
``` yaml
services:
  - class: org.yamcs.http.HttpServer
  - class: org.yamcs.ProcessRunner
    args: 
      command: "./python_scripts/test.py"
      restart: "always"

```

i get this error: 
![image](https://user-images.githubusercontent.com/46066434/169802062-8a813942-d389-4473-b1bd-5e14490c56d7.png)
How can i solve this?

**Reply by fqqb** [reply]
1/ Will try later myself, but I suspect it can be fixed by somehow enabling PYTHONUNBUFFERED (https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUNBUFFERED)
2/ The option was added in 5.6.2, check your Yamcs dependency is up-to-date.

**Reply by bruno-luca** [reply]
Thanks, both of your tips worked for me! For anyone else who is using `ProcessRunner` to launch a python script with `shebang` line this is how to enable PYTHONUNBUFFERED:

```python
#!/usr/bin/env -S python3 -u

import  socket

...
```

**Comment by xpromache** [comment]
yes, indeed, you can just copy the FilePollingTmDataLink.

Maybe we should add an option in that class (incomingFile as alternative to
the current incomingDir) rather than making another link.

If you feel up to it, I will gladly review and accept a pull request.

nicolae
